dali/sasha/saarch.cpp

```diff
@@ -83,297 +83,449 @@ static void mkDateCompare(bool dfu,const char *dt,StringBuffer &out,char fill)
+        }
+        bool addOutputFromWUTree(StringArray &outputFields, IPropertyTree *wuTree)
+        {
+            if (outputXML)
+            { //cmd->getAction()==SCA_GET
+                StringBuffer xml;
+                toXML(wuTree, xml);
+                if (!cmd->addResult(xml.str()))
+                    return false; //Log overflowed?
+            }
+            else
+            {
+                if (outputFields.length() == 0)
+                    cmd->addId(wuTree->queryName());
+                else
+                {
+                    StringBuffer output = wuTree->queryName();
+                    //Append more into the output.
+                    setWUDataTree(wuTree, outputFields, output);
+                    cmd->addId(output.str());
                 }
             }
+            return true;
         }
-    }
-    if (cmd->getOnline()) {
-        if (haswusoutput)
-            throw MakeStringException(-1,"SCA_WORKUNIT_SERVICES_GET not implemented for online workunits!");
-        StringBuffer xpath(baseXPath);
-        if (!conn)
+        void setWUDataTree(IPropertyTree *wu, StringArray &outputFields, StringBuffer &output)
         {
-            if (!isWild)
-                xpath.append("/").append(mask);
-            conn.setown(querySDS().connect(xpath.str(), myProcessSession(), 0, 5*60*1000));
+            ForEachItemIn(i, outputFields)
+            {
+                const char* outputField = (const char*)outputFields.item(i);
+
+                StringBuffer val;
+                bool found = true;
+                if (strieq(outputField, "owner"))
+                    wu->getProp("@submitID", val);
+                else if (strieq(outputField, "cluster"))
+                    wu->getProp("@clusterName", val);
+                else if (strieq(outputField, "jobname"))
+                    wu->getProp("@jobName",val);
+                else if (strieq(outputField,"state"))
+                    wu->getProp("@state", val);
+                else
+                    found = false;
+                if (found)
+                    output.append(',').append(val);
+            }
         }
-        if (conn)
+        void addWUsSorted(StringArray &outputFields, IArrayOf<IPropertyTree> &wus)
         {
-            Owned<IPropertyTreeIterator> iter = conn->queryRoot()->getElements(isWild ? "*" : NULL);
-            unsigned index = 0;
-            StringBuffer val;
-            ForEach(*iter) {
-                IPropertyTree &pt=iter->query();
-                const char *wuid = pt.queryName();
-                if (index>start+num)
+            unsigned nwus = wus.ordinality();
+            unsigned *idx = new unsigned[nwus];
+            unsigned i;
+            for (i = 0; i < nwus; i++)
+                idx[i] = i;
+            {
+                CriticalBlock block(wuDataSortSect);
+                wusRef = &wus;
+                wuDataSortInc = sortInc;
+                qsort(idx, nwus, sizeof(unsigned), compare);
```

Comment:
globals and locks should be necessary.
You should be able to pass a sort compare function into the array::sort method directly.

Reply:
Will use array::sort.